packages/kbn-dev-utils/src/ci_stats_reporter/ci_stats_reporter.ts

```diff
@@ -7,106 +7,187 @@
+export interface CiStatsTimingMetadata {
+  [key: string]: string | string[] | number | boolean | undefined;
+}
+export interface CiStatsTiming {
+  group: string;
+  id: string;
+  ms: number;
+  meta?: CiStatsTimingMetadata;
+}
 
-  let config: unknown;
-  try {
-    config = JSON.parse(configJson);
-  } catch (_) {
-    // handled below
-  }
+export interface ReqOptions {
+  auth: boolean;
+  path: string;
+  body: any;
+  bodyDesc: string;
+}
 
-  if (typeof config === 'object' && config !== null) {
-    return validateConfig(log, config as { [k in keyof Config]: unknown });
+export interface TimingsOptions {
+  /** list of timings to record */
+  timings: CiStatsTiming[];
+  /** master, 7.x, etc, automatically detected from package.json if not specified */
+  upstreamBranch?: string;
+  /** value of data/uuid, automatically loaded if not specified */
+  kibanaUuid?: string | null;
+}
+export class CiStatsReporter {
+  static fromEnv(log: ToolingLog) {
+    return new CiStatsReporter(parseConfig(log), log);
   }
 
-  log.warning('KIBANA_CI_STATS_CONFIG is invalid, stats will not be reported');
-  return;
-}
+  constructor(private config: Config | undefined, private log: ToolingLog) {}
 
-function validateConfig(log: ToolingLog, config: { [k in keyof Config]: unknown }) {
-  const validApiUrl = typeof config.apiUrl === 'string' && config.apiUrl.length !== 0;
-  if (!validApiUrl) {
-    log.warning('KIBANA_CI_STATS_CONFIG is missing a valid api url, stats will not be reported');
-    return;
+  isEnabled() {
+    return process.env.CI_STATS_DISABLED !== 'true';
   }
 
-  const validApiToken = typeof config.apiToken === 'string' && config.apiToken.length !== 0;
-  if (!validApiToken) {
-    log.warning('KIBANA_CI_STATS_CONFIG is missing a valid api token, stats will not be reported');
-    return;
+  hasBuildConfig() {
+    return this.isEnabled() && !!this.config?.apiToken && !!this.config?.buildId;
   }
 
-  const validId = typeof config.buildId === 'string' && config.buildId.length !== 0;
-  if (!validId) {
-    log.warning('KIBANA_CI_STATS_CONFIG is missing a valid build id, stats will not be reported');
-    return;
+  /**
+   * Report timings data to the ci-stats service. If running in CI then the reporter
+   * will include the buildId in the report with the access token, otherwise the timings
+   * data will be recorded as anonymous timing data.
+   */
+  async timings(options: TimingsOptions) {
+    if (!this.isEnabled()) {
+      return;
+    }
+
+    const buildId = this.config?.buildId;
+    const timings = options.timings;
+    const upstreamBranch = options.upstreamBranch ?? this.getUpstreamBranch();
+    const kibanaUuid = options.kibanaUuid === undefined ? this.getKibanaUuid() : options.kibanaUuid;
+    const defaultMetadata = {
+      osPlatform: Os.platform(),
+      osRelease: Os.release(),
+      osArch: Os.arch(),
+      cpuCount: Os.cpus()?.length,
+      cpuModel: Os.cpus()[0]?.model,
+      cpuSpeed: Os.cpus()[0]?.speed,
```

Comment:
Thoughts on also collecting `Os.freemem()` and/or `Os.totalmem()`

Reply:
I think `Os.totalmom()` might be useful, and maybe some form of generalization around "memory available", like if "verylittle", "some", and "plenty"?

Reply:
Thinking through this we probably just would like to know when freemem is below 4GB or something, so maybe it makes sense to just index the freemem too. I'm on board.

Reply:
What would be the reason not to include the raw numbers? That way we could create a dynamic field for some threshold, like 4G or what we would determine to be problematic.

Reply:
Yeah, that's what I'm thinking, we should just store the numbers and decide later if we want to create bands or something.